### PR TITLE
add dssm model

### PR DESCRIPTION
**Pull request by bwanglzu**
for issue #138 

Adapt dssm to new `base_model`.

### CODE REVIEW

File: matchzoo/models/dssm_model.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    """
+
+    @classmethod
+    def get_default_params(cls) -> engine.ParamTable:
+        """:return: model default parameters."""
+        params = super().get_default_params()
+        params.add(engine.Param('w_initializer', 'glorot_normal'))
+        params.add(engine.Param('b_initializer', 'zeros'))
+        # TODO GET TRI-LETTER DIMENSIONALITY FROM FIT-TRANSFORM AS INPUT SHAPE
+        params.add(engine.Param('input_shapes', [(30000,), (30000,)]))
```
Comment (uduse) [inline matchzoo/models/dssm_model.py]: Should do `params['input_shapes'] = ...` instead. I'll add a checking for already added params in a PR later.

File: matchzoo/models/dssm_model.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        params.add(engine.Param('dim_fan_out', 128))
+        params.add(engine.Param('dim_hidden', 300))
+        params.add(engine.Param('activation_hidden', 'tanh'))
+        params.add(engine.Param('activation_prediction', 'softmax'))
+        params.add(engine.Param('lr', 0.01))
```
Comment (uduse) [inline matchzoo/models/dssm_model.py]: I am thinking about relaxing the constraint of "using only native data types or matchzoo data types" for params. This way, learning rate is no longer a param itself. It is and it should be one of the parameters of `params['optimizer']`. So when the user really wants to tweaking learning rate, the user can do `params['optimizer'] = keras.optimizers.SGD(learning_rate=0.01)` or something. It's not friendly to parameter tuning though, but I'll figure that out separately.

File: matchzoo/models/dssm_model.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        params.add(engine.Param('decay', 1e-6))
```
Comment (uduse) [inline matchzoo/models/dssm_model.py]: Same as `lr`.

File: matchzoo/models/dssm_model.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        params.add(engine.Param('batch_size', 1024))
+        params.add(engine.Param('num_epochs', 20))
```
Comment (uduse) [inline matchzoo/models/dssm_model.py]: `batch_size` and `num_epochs` are not part of model structures. We should set them manually for now and add them into a higher level API to models (e.g. something called a `Trainer`) later.

File: matchzoo/models/dssm_model.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        params.add(engine.Param('train_test_split', 0.2))
+        params.add(engine.Param('verbose', 1))
```
Comment (uduse) [inline matchzoo/models/dssm_model.py]: `verbose`? Not needed for now. We can do all the loggers later and make `verbose` a shared parameter in the base model.

File: matchzoo/models/dssm_model.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        params.add(engine.Param('shuffle', True))
```
Comment (uduse) [inline matchzoo/models/dssm_model.py]: Also, not a part of the model structure.

File: matchzoo/models/dssm_model.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        params.add(engine.Param('optimizer', 'sgd'))
+        params.add(engine.Param('num_hidden_layers', 2))
+        params.add(engine.Param('loss', 'categorical_crossentropy'))
```
Comment (uduse) [inline matchzoo/models/dssm_model.py]: It's based on `params['task']`.

File: matchzoo/models/dssm_model.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        params.add(engine.Param('activation_hidden', 'tanh'))
+        params.add(engine.Param('activation_prediction', 'softmax'))
+        params.add(engine.Param('lr', 0.01))
+        params.add(engine.Param('decay', 1e-6))
+        params.add(engine.Param('batch_size', 1024))
+        params.add(engine.Param('num_epochs', 20))
+        params.add(engine.Param('train_test_split', 0.2))
+        params.add(engine.Param('verbose', 1))
+        params.add(engine.Param('shuffle', True))
+        params.add(engine.Param('optimizer', 'sgd'))
```
Comment (uduse) [inline matchzoo/models/dssm_model.py]: Use `params['optimizer'] = 'sgd'`.

File: tests/unit_test/models/test_dssm_model.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    assert model.predict(x) is not None
+
+
+def test_save_load_model(model):
+    tmpdir = '.tmpdir'
+    model.save(tmpdir)
+    assert engine.load_model(tmpdir)
+    with pytest.raises(FileExistsError):
+        model.save(tmpdir)
+    shutil.rmtree(tmpdir)
```
Comment (uduse) [inline tests/unit_test/models/test_dssm_model.py]: I don't think each model needs its own tests. A simple doctest should give 100% coverage already.

File: matchzoo/models/dssm_model.py (hunk 1 above)
Comment (bwanglzu) [reply matchzoo/models/dssm_model.py]: when to use `params setter`, and when to use `engine.Param`?

File: matchzoo/models/dssm_model.py (hunk 5 above)
Comment (bwanglzu) [reply matchzoo/models/dssm_model.py]: that's better

File: matchzoo/models/dssm_model.py (hunk 1 above)
Comment (bwanglzu) [reply matchzoo/models/dssm_model.py]: update I get the idea

File: matchzoo/models/dssm_model.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
+"""An implementation of DSSM, Deep Structured Semantic Model."""
+from matchzoo import engine
+from keras.models import Model
+from keras.layers import Dense, Input, Dot
+
+
+class DssmModel(engine.BaseModel):
```
[the remark this answers is not in the document]
Comment (bwanglzu) [reply matchzoo/models/dssm_model.py]: done

File: matchzoo/models/dssm_model.py (hunk 2 above)
Comment (bwanglzu) [reply matchzoo/models/dssm_model.py]: ok let's discuss later

File: matchzoo/models/dssm_model.py (hunk 3 above)
Comment (bwanglzu) [reply matchzoo/models/dssm_model.py]: done

File: tests/unit_test/models/test_dssm_model.py (hunk 9 above)
Comment (bwanglzu) [reply tests/unit_test/models/test_dssm_model.py]: if we remove this, the coverage check is going to fail

File: matchzoo/models/dssm_model.py (hunk 8 above)
Comment (bwanglzu) [reply matchzoo/models/dssm_model.py]: done

File: matchzoo/models/dssm_model.py (hunk 7 above)
Comment (bwanglzu) [reply matchzoo/models/dssm_model.py]: do something like this?
```python
params['task'] = 'classification'
params['loss']='categorical_crossentropy'
```

Comment (bwanglzu) [reply matchzoo/models/dssm_model.py]: that's correct

File: tests/unit_test/models/test_dssm_model.py (hunk 9 above)
Comment (uduse) [reply tests/unit_test/models/test_dssm_model.py]: #191 I just added standardized model testing procedure in this PR. Testing a new model with one line.

Comment (bwanglzu) [reply tests/unit_test/models/test_dssm_model.py]: okay @uduse wait unitl #191  merged.

File: matchzoo/models/dssm_model.py [hunk 11]
```diff
@@ -39,44 +38,51 @@ def get_default_params(cls) -> engine.ParamTable:
...
+        x = [base_network(input_left),
+             base_network(input_right)]
         # Dot product with cosine similarity.
-        x = Dot(axes=1,
-                normalize=True)(x_in)
-        x_out = Activation(activation=self._params['activation_prediction'])(x)
-        self._backend = Model(inputs=x_in, outputs=x_out)
+        x = Dot(axes=[1, 1],
+                normalize=True)(x)
+        x_out = Dense(2,
```
[the remark this answers is not in the document]
Comment (bwanglzu) [reply matchzoo/models/dssm_model.py]: @faneshion Then for the ranking case, we need to sue `sigmoid` instead of `softmax` to transform the result to prob, or write our activation.

File: matchzoo/models/dssm_model.py [hunk 12]
```diff
@@ -39,44 +38,51 @@ def get_default_params(cls) -> engine.ParamTable:
...
 
-        :return: 128 dimension vector representation.
+        :return: x: 128d vector(tensor) representation.
         """
-        x_in = Input(shape=(dim_triletter,))
-        # Initialize input layer.
-        x = Dense(units=self._params['dim_hidden'],
-                  input_shape=(dim_triletter,),
+        input = Input(shape=input_shape)
+        x = Dense(self._params['dim_hidden'],
```
[the remark this answers is not in the document]
Comment (bwanglzu) [reply matchzoo/models/dssm_model.py]: interesting, let's discuss it. might affect varies models.

[the remark this answers is not in the document]
Comment (bwanglzu) [reply matchzoo/models/dssm_model.py]: @faneshion I've checked the keras API, the `sparse` argument is not available at the moment (it's there, but might cause bugs). Let's use normal Input layer at the moment, and refactor in the future.
